File: 2_ReactWeb/src/features/conversation-export/model/conversationExport.ts
export type ConversationExportFormat = "docx" | "markdown" | "txt" | "html" | "json";

export type ConversationExportRange =
  | "conversation"
  | "message"
  | "through-message"
  | "from-message";

export type ConversationExportContentKey =
  | "sessionInfo"
  | "assistantContent"
  | "userMessages"
  | "thinking"
  | "toolCalls"
  | "toolResults"
  | "errorMessages"
  | "systemMessages"
  | "timestamps"
  | "images"
  | "modelInfo"
  | "tokenUsage"
  | "messageMetadata";

export type ConversationExportContentSelection = Record<
  ConversationExportContentKey,
  boolean
>;

export type ConversationExportRequest = {
  initialDirectory: string;
  messageId: string | null;
  projectId: string;
  scope: "conversation" | "message";
  sessionId: string;
  sessionTitle: string;
};

export const CONVERSATION_EXPORT_FORMATS: ReadonlyArray<{
  extension: string;
  format: ConversationExportFormat;
}> = [
  { format: "docx", extension: ".docx" },
  { format: "markdown", extension: ".md" },
  { format: "txt", extension: ".txt" },
  { format: "html", extension: ".html" },
  { format: "json", extension: ".json" },
];

const COMMON_CONTENT_OPTIONS: ConversationExportContentKey[] = [
  "sessionInfo",
  "assistantContent",
  "userMessages",
  "thinking",
  "toolCalls",
  "toolResults",
  "errorMessages",
  "systemMessages",
  "timestamps",
  "modelInfo",
  "tokenUsage",
];

export const DEFAULT_CONVERSATION_EXPORT_CONTENT: ConversationExportContentSelection = {
  sessionInfo: true,
  assistantContent: true,
  userMessages: true,
  thinking: false,
  toolCalls: false,
  toolResults: false,
  errorMessages: true,
  systemMessages: false,
  timestamps: false,
  images: true,
  modelInfo: false,
  tokenUsage: false,
  messageMetadata: false,
};

export function getConversationExportContentOptions(
  format: ConversationExportFormat,
): ConversationExportContentKey[] {
  if (format === "json") {
    return [...COMMON_CONTENT_OPTIONS, "messageMetadata"];
  }
  if (format === "txt") {
    return COMMON_CONTENT_OPTIONS;
  }
  return [...COMMON_CONTENT_OPTIONS, "images"];
}

export function getDefaultConversationExportRange(
  request: ConversationExportRequest,
): ConversationExportRange {
  return request.scope === "conversation" ? "conversation" : "message";
}

export function buildConversationExportBaseName(
  request: ConversationExportRequest,
  now = new Date(),
) {
  const normalizedTitle = (request.sessionTitle.trim() || "会话")
    .replace(/[<>:"/\\|?*\u0000-\u001f]/g, "_")
    .replace(/[. ]+$/g, "")
    .replace(/_+$/g, "")
    .trim();
  return `${normalizedTitle || "会话"}_${formatExportTimestamp(now)}`;
}

function formatExportTimestamp(value: Date) {
  const date = [
    value.getFullYear(),
    padDatePart(value.getMonth() + 1),
    padDatePart(value.getDate()),
  ].join("-");
  const time = [
    padDatePart(value.getHours()),
    padDatePart(value.getMinutes()),
    padDatePart(value.getSeconds()),
  ].join("-");
  return `${date}_${time}`;
}

function padDatePart(value: number) {
  return String(value).padStart(2, "0");
}

File: 1_PythonServer/app/domain/project/conversation_export.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from enum import StrEnum
from pathlib import Path

from app.domain.project.project_conversation import (
    ProjectConversationMessage,
    ProjectConversationSession,
)


class ConversationExportFormat(StrEnum):
    DOCX = "docx"
    MARKDOWN = "markdown"
    TXT = "txt"
    HTML = "html"
    JSON = "json"


class ConversationExportRange(StrEnum):
    CONVERSATION = "conversation"
    MESSAGE = "message"
    THROUGH_MESSAGE = "through-message"
    FROM_MESSAGE = "from-message"


@dataclass(frozen=True, slots=True)
class ConversationExportContentSelection:
    session_info: bool
    assistant_content: bool
    user_messages: bool
    thinking: bool
    tool_calls: bool
    tool_results: bool
    error_messages: bool
    system_messages: bool
    timestamps: bool
    images: bool
    model_info: bool
    token_usage: bool
    message_metadata: bool

    def has_supported_content(self, export_format: ConversationExportFormat) -> bool:
        values = [
            self.session_info,
            self.assistant_content,
            self.user_messages,
            self.thinking,
            self.tool_calls,
            self.tool_results,
            self.error_messages,
            self.system_messages,
            self.timestamps,
            self.model_info,
            self.token_usage,
        ]
        if export_format not in {ConversationExportFormat.TXT, ConversationExportFormat.JSON}:
            values.append(self.images)
        if export_format == ConversationExportFormat.JSON:
            values.append(self.message_metadata)
        return any(values)


@dataclass(frozen=True, slots=True)
class ConversationExportDocument:
    project_root: Path
    session: ProjectConversationSession
    messages: tuple[ProjectConversationMessage, ...]
    export_range: ConversationExportRange
    exported_at: datetime


@dataclass(frozen=True, slots=True)
class ConversationExportImage:
    asset_name: str
    alt_text: str
    content: bytes
    embedded: bool
    message_id: str
    mime_type: str
    source: str


@dataclass(frozen=True, slots=True)
class PreparedConversationExport:
    document: ConversationExportDocument
    images: tuple[ConversationExportImage, ...]

    def images_for_message(self, message_id: str) -> tuple[ConversationExportImage, ...]:
        return tuple(image for image in self.images if image.message_id == message_id)


@dataclass(frozen=True, slots=True)
class ConversationExportFile:
    relative_path: str
    content: bytes


@dataclass(frozen=True, slots=True)
class RenderedConversationExport:
    content: bytes
    extension: str
    bundle: bool = False
    files: tuple[ConversationExportFile, ...] = ()
    warnings: tuple[str, ...] = ()


@dataclass(frozen=True, slots=True)
class StoredConversationExport:
    container_path: Path
    output_path: Path


@dataclass(frozen=True, slots=True)
class ConversationExportResult:
    container_path: Path
    output_path: Path
    message_count: int
    warnings: tuple[str, ...]
